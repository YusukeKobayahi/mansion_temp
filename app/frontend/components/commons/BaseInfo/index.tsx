import styles from "./index.module.scss";
import UnderlineHeading from "~/components/commons/headings/UnderlineHeading";
import classnames from "classnames";
export interface BaseInfoProps {
  mainText: string;
  areaText: string;
  isNoindex: boolean;
  totalCount: number;
}

const BaseInfo: React.FC<BaseInfoProps> = ({
  mainText,
  areaText,
  isNoindex,
  totalCount,
}: BaseInfoProps) => {
  const explanation =
    totalCount == 0
      ? "条件にあう物件がありません。条件を変更して再度検索してください。"
      : `Housiiでは、「${
          isNoindex ? mainText + "のあなたの条件にあった」" : mainText + "」の"
        }中古マンションについて紹介しています。`;
  return (
    <div>
      <h1 className={styles.title}>
        <UnderlineHeading>
          <p className={"pcOnly"}>
            {isNoindex ? mainText + "のあなたの条件にあった" : mainText + "の"}
            中古マンション一覧
          </p>
          <p className={"spOnly"}>{areaText}の中古マンション一覧</p>
        </UnderlineHeading>
      </h1>
      <p className={classnames(styles.info, "pcOnly")}>{explanation}</p>
    </div>
  );
};

export default BaseInfo;
