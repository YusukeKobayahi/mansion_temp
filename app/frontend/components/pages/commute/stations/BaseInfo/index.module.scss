@import "~/styles/utils";

.title {
  padding-top: 20px;
}

.info {
  font-size: 16px;
  color: #746744;
  line-height: 1.5;

  @include screenMd {
    padding: 0 16px;
    font-size: 14px;
  }
}
