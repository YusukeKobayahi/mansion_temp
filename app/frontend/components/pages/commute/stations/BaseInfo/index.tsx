import styles from "./index.module.scss";
import UnderlineHeading from "~/components/commons/headings/UnderlineHeading";

type BaseInfoProps = {
  mainText: string;
};

const BaseInfo: React.FC<BaseInfoProps> = ({ mainText }: BaseInfoProps) => {
  return (
    <div>
      <h1 className={styles.title}>
        <UnderlineHeading>
          <p>{mainText}</p>
        </UnderlineHeading>
      </h1>
    </div>
  );
};

export default BaseInfo;
