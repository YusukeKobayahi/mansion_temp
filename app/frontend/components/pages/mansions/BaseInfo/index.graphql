fragment MsLibraryMansionsBaseInfo on Mansion {
  uniqueCode
  name
  prefecture {
    name
    jisCode
  }
  city {
    name
    jisCode
  }
  town {
    id
    name
  }
  street {
    name
  }
  age
  constructedIn
  blockNumber
  access {
    id
    stationId
    stationLineName
    stationName
    stationWalkingMinutes
    lineId
  }
  display {
    minExclusiveArea
    maxExclusiveArea
  }
}
