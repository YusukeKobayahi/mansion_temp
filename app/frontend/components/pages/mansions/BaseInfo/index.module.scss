@import "~/styles/variables";
@import "~/styles/utils";

.container {
  position: relative;
  margin-top: 64px;

  @include screenMd {
    margin-top: 25px;
  }
}

.info {
  margin-top: 25px;
  font-size: 16px;
  color: $color-main;
  line-height: 1.5;

  @include screenMd {
    text-align: left;
    margin-top: 16px;
    padding: 0 16px;
    font-size: 14px;
  }

  button {
    text-decoration: underline;
    padding: 0;
    background-color: unset;
    color: $color-main;
  }
}
