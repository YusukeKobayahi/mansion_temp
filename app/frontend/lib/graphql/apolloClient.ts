// 参考
// https://github.com/vercel/next.js/tree/canary/examples/with-apollo

import { useMemo } from "react";
import { ApolloClient, ApolloLink, InMemoryCache } from "@apollo/client";
import { createHttpLink } from "apollo-link-http";
import { from } from "apollo-link";
import { createFragmentArgumentLink } from "apollo-link-fragment-argument";
import { mergeDeep, relayStylePagination } from "@apollo/client/utilities";
import { isSSR } from "~/lib/utils";

/* eslint-disable @typescript-eslint/no-explicit-any -- キャッシュの型を事前に決められないため */
type CacheState = Record<string, any> | null;
/* eslint-enable @typescript-eslint/no-explicit-any */

/* eslint-disable @typescript-eslint/naming-convention -- GraphQLのタイプ名 (このconventionに従わない)を指定したいので */
const createCache = () =>
  new InMemoryCache({
    typePolicies: {
      Mansion: {
        keyFields: ["uniqueCode"],
        fields: {
          salesHistories: relayStylePagination(),
        },
      },
    },
  });
/* eslint-enable @typescript-eslint/naming-convention */

const link = from([
  createFragmentArgumentLink(),
  createHttpLink({
    uri: `${process.env.NEXT_PUBLIC_CORE_API_URL}/consumer/graphql`,
    fetchOptions: {
      mode: "cors",
    },
  }),
]);

function createApolloClient() {
  return new ApolloClient({
    ssrMode: isSSR(),
    // See: https://github.com/apollographql/apollo-link/issues/1258#issuecomment-604031488
    // TODO: @apollo/clientバージョン上げたら直るかも
    link: (link as unknown) as ApolloLink,
    cache: createCache(),
    defaultOptions: {
      watchQuery: {
        // SSR時にuseQueryによってリクエストしてレンダリングする場合、
        // getServerSidePropsからキャッシュが渡っている必要がある。(でないと描画されない)
        // そのため、SSR時は常にcacheから取得されるべきで、実装意図を明確にするためPolicyも設定しておく
        // (エラーになるわけではない)
        fetchPolicy: isSSR() ? "cache-only" : "cache-first",
      },
    },
  });
}

let apolloClient: ReturnType<typeof createApolloClient>;

export function initializeApollo(
  initialState: CacheState = null
): ApolloClient<CacheState> {
  const _apolloClient = apolloClient ?? createApolloClient();

  // If your page has Next.js data fetching methods that use Apollo Client, the initial state
  // gets hydrated here
  if (initialState) {
    // Get existing cache, loaded during client side data fetching
    const existingCache = _apolloClient.extract();

    // Merge the existing cache into data passed from getStaticProps/getServerSideProps
    const data = mergeDeep(initialState, existingCache);

    // Restore the cache with the merged data
    _apolloClient.cache.restore(data);
  }
  // For SSG and SSR always create a new Apollo Client
  if (isSSR()) return _apolloClient;
  // Create the Apollo Client once in the client
  if (!apolloClient) apolloClient = _apolloClient;

  return _apolloClient;
}

export function useApollo(initialState: CacheState): ApolloClient<CacheState> {
  const store = useMemo(() => initializeApollo(initialState), [initialState]);
  return store;
}
