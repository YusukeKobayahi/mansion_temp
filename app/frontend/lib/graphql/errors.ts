export interface GraphQLError {
  graphQLErrors?: {
    message: string;
    extensions?: {
      code: string;
    };
  }[];
}

export const isNotFoundError = (err: GraphQLError): boolean => {
  const graphQLErrors = err.graphQLErrors;
  if (graphQLErrors == null) return false;

  return graphQLErrors[0]?.extensions?.code === "NOT_FOUND";
};
