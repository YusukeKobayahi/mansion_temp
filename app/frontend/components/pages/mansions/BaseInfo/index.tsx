import Link from "next/link";
import { buildLibraryPagePath, buildConstructedDate } from "~/lib/utils";
import { event } from "~/lib/gtag";
import compact from "lodash/compact";
import { useMemo } from "react";
import { useDeepCompareMemo } from "use-deep-compare";
import styles from "./index.module.scss";
import { MsLibraryMansionsBaseInfoFragment } from "~/graphql/generated";
import UnderlineHeading from "~/components/commons/headings/UnderlineHeading";
import { GoogleAnalyticsEvent } from "~/lib/types";

export interface BaseInfoProps {
  data: MsLibraryMansionsBaseInfoFragment;
}

const BaseInfo: React.FC<BaseInfoProps> = ({ data }: BaseInfoProps) => {
  const clickArea = (label: GoogleAnalyticsEvent["label"]) => {
    event({
      action: "msl_area_click",
      category: "msl_shosai",
      label: label,
      value: 1,
    });
  };
  const constructedDate = buildConstructedDate(data.constructedIn, data.age);
  const address = useDeepCompareMemo(() => {
    return (
      <span>
        <Link
          href={"/prefectures/[jisCode]"}
          as={buildLibraryPagePath("prefectures", data.prefecture.jisCode)}
          prefetch={false}
          passHref
        >
          <button className={styles.link} onClick={() => clickArea("msl_pref")}>
            {data.prefecture.name}
          </button>
        </Link>
        &nbsp;
        <Link href={buildLibraryPagePath("cities", data.city.jisCode)} passHref>
          <button className={styles.link} onClick={() => clickArea("msl_city")}>
            {data.city.name}
          </button>
        </Link>
        &nbsp;
        <Link
          href={{
            pathname: buildLibraryPagePath("cities", data.city.jisCode),
            query: { tw: data.town.id },
          }}
          passHref
        >
          <button onClick={() => clickArea("msl_town")} className={styles.link}>
            {data.town.name}
          </button>
        </Link>
        &nbsp;
        {compact([data.street?.name, data.blockNumber]).join("")}
      </span>
    );
  }, [
    data.prefecture.name,
    data.city.name,
    data.town.name,
    data.street?.name,
    data.blockNumber,
  ]);
  const accessInfo = useDeepCompareMemo(() => {
    return data.access.map((a, i) => {
      return (
        <span key={i}>
          <Link href={buildLibraryPagePath("lines", String(a.lineId))} passHref>
            <button onClick={() => clickArea("msl_line")}>
              {a.stationLineName}
            </button>
          </Link>
          「
          <Link href={buildLibraryPagePath("stations", String(a.stationId))}>
            <button onClick={() => clickArea("msl_station")}>
              {a.stationName}
            </button>
          </Link>
          」
          {a.stationWalkingMinutes != null &&
            "徒歩" + a.stationWalkingMinutes + "分"}
        </span>
      );
    });
  }, [data.access]);

  const exclusiveArea = useMemo(() => {
    if (data.display.maxExclusiveArea === data.display.minExclusiveArea) {
      return `${data.display.maxExclusiveArea}㎡`;
    } else {
      return `${data.display.minExclusiveArea}㎡~${data.display.maxExclusiveArea}㎡`;
    }
  }, [data.display.minExclusiveArea, data.display.maxExclusiveArea]);

  return (
    <section className={styles.container}>
      <div className={styles.head}>
        <h1>
          <UnderlineHeading>{data.name}</UnderlineHeading>
        </h1>
      </div>
      <div className={styles.info}>
        <p>
          <span>占有面積 {exclusiveArea} / </span>
          <span>築年月 {constructedDate} / </span>
          <span>最寄駅 {accessInfo}</span>
        </p>
        <p>
          <span>所在地 {address}</span>
        </p>
      </div>
    </section>
  );
};

export default BaseInfo;
