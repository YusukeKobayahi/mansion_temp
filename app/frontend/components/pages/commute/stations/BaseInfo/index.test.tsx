import { render } from "@testing-library/react";
import mockData from "../fixtures/mockData.json";
import { CommuteStationsQuery } from "~/graphql/generated";
import BaseInfo from ".";

describe("components/pages/commute/stations/BaseInfo", () => {
  const mock = (mockData as unknown) as CommuteStationsQuery;

  it("正常に表示される", () => {
    const { asFragment } = render(<BaseInfo mainText={mock.station.name} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
