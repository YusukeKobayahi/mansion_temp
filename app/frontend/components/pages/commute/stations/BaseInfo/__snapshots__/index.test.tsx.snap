// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/pages/commute/stations/BaseInfo 正常に表示される 1`] = `
<DocumentFragment>
  <div>
    <h1
      class="title"
    >
      <div
        class="heading"
      >
        <span>
          <p>
            品川
          </p>
        </span>
      </div>
    </h1>
  </div>
</DocumentFragment>
`;
